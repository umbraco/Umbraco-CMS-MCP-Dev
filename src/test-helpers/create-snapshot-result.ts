import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
