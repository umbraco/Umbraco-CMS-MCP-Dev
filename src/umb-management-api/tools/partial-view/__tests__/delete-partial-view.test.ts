import DeletePartialViewTool from "../delete/delete-partial-view.js";
import { PartialViewBuilder } from "./helpers/partial-view-builder.js";
import { PartialViewHelper } from "./helpers/partial-view-helper.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";

const TEST_PARTIAL_VIEW_NAME = "_TestDeletePartialView.cshtml";
const TEST_CONTENT = "@* Test delete content *@\n<div><p>Delete Test Content</p></div>";
const NON_EXISTENT_PATH = "/_NonExistentPartialView.cshtml";

describe("delete-partial-view", () => {
  let originalConsoleError: typeof console.error;
  let builder: PartialViewBuilder;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
    builder = new PartialViewBuilder();
  });

  afterEach(async () => {
    console.error = originalConsoleError;
  });

  it("should delete a partial view", async () => {
    // Arrange - Create partial view to delete
    await builder
      .withName(TEST_PARTIAL_VIEW_NAME)
      .withContent(TEST_CONTENT)
      .create();

    // Verify it exists before deletion
    const existsBeforeDelete = await PartialViewHelper.verifyPartialView(builder.getPath());
    expect(existsBeforeDelete).toBe(true);

    const params = {
      path: builder.getPath()
    };

    // Act
    const result = await DeletePartialViewTool().handler(params, { signal: new AbortController().signal });

    // Assert
    const normalizedResult = createSnapshotResult(result);
    expect(normalizedResult).toMatchSnapshot();

    // Verify the partial view no longer exists
    const existsAfterDelete = await PartialViewHelper.verifyPartialView(builder.getPath());
    expect(existsAfterDelete).toBe(false);
  });

  it("should handle non-existent partial view", async () => {
    // Arrange
    const params = {
      path: NON_EXISTENT_PATH
    };

    // Act
    const result = await DeletePartialViewTool().handler(params, { signal: new AbortController().signal });

    // Assert - Error responses don't use createSnapshotResult
    expect(result).toMatchSnapshot();
  });

});