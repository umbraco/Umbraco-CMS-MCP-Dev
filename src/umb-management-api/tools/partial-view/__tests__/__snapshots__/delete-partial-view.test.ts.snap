// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-partial-view should delete a partial view 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`delete-partial-view should handle non-existent partial view 1`] = `
{
  "content": [
    {
      "text": "Error using delete-partial-view:
{
  "message": "Request failed with status code 404",
  "response": {
    "type": "Error",
    "title": "Partial view not found",
    "status": 404,
    "detail": "The partial view was not found.",
    "operationStatus": "NotFound"
  }
}",
      "type": "text",
    },
  ],
}
`;
